README.md
# LittleGameEngine
Un pequeño (e ineficiente) motor de juegos para aprender a programar con Python

```python
from lge.LittleGameEngine import LittleGameEngine
from lge.Sprite import Sprite
from lge.Canvas import Canvas

# creamos el juego
winSize = (800, 440)
lge = LittleGameEngine(winSize, "The World", (255, 255, 0))

# cargamos los recursos que usaremos
resourceDir = "../resources"

lge.loadImage("fondo", resourceDir + "/images/Backgrounds/FreeTileset/Fondo.png", winSize)
lge.loadImage("heroe", resourceDir + "/images/Swordsman/Idle/Idle_0*.png", 0.08)
lge.loadTTFFont("backlash.40", resourceDir + "/fonts/backlash.ttf", 40)
lge.loadSound("fondo", resourceDir + "/sounds/happy-and-sad.wav")

# activamos la musica de fondo
lge.playSound("fondo", True, 50)

# agregamos el fondo
fondo = Sprite("fondo", (0, 0))
lge.addGObject(fondo, 0)

# agregamos al heroe
heroe = Sprite("heroe", (226, 142))
lge.addGObject(heroe, 1)

# agregamos un texto con transparencia
canvas = Canvas((200, 110), (400, 200))
canvas.drawText("Little Game Engine", (30, 90), "backlash.40", (20, 20, 20))
lge.addGObjectGUI(canvas)

# main loop
lge.run(60)
```
![](images/world.png)
![](images/collisions.png)

![](images/Betty.png)
![](images/Plataforma.png)

![](images/Pong.png)
![](images/Canvas.png)

![](images/Birds.png)


## Imágenes
- https://opengameart.org/content/one-more-lpc-alternate-character
- https://opengameart.org/content/free-platformer-game-tileset
- https://opengameart.org/content/2d-game-character-pack-slim-version
- https://opengameart.org/content/game-character-blue-flappy-bird-sprite-sheets
- https://opengameart.org/content/dungeon-crawl-32x32-tiles
- https://www.kenney.nl
- https://opengameart.org/content/2d-platformer-volcano-pack-11

## Sonidos
- https://freesound.org/people/TiagoThanos/sounds/571229/
- https://freesound.org/people/MATRIXXX_/sounds/365668/

## Fonts
- backlash.ttf - Backlash BRK - No contiene información de licencia
- FreeMono.ttf - FreeMono - Parte de los FreeFont de GNU - GNU General Public License
- LiberationMonoRegular.ttf - Liberation Mono - Liberation Fonts License


CHANGELOG.md
# ChangeLog

# v0.7.1 2022-06-23
- Renombre método para carga de imágenes

# v0.7.0 2022-06-22
- Agrega método LittleGameEngine.contains() para obtener todos los GameObjects que contienen a un punto dado

# 2022-06-22
- Crea clase ImageManager para el manejo de imagen
- Crea clase SoundManager para el manejo de sonidos
- Crea clase Fontmanager para el manejo de fonts
- Manejo de imágenes, sonido y fonts se llevan a las clases señaladas
- Canvas agrega método para mostrar imágenes precargadas
- Se corrigen los demos acorde a lo anterior
- Se reestructuran directorios. Todo quedaen en src/ y demos/ cambia de nombre a test/

# v0.6.0 2022-05-28
- Versión estable

# v0.5.7.2 2022-05-28
- Renombra loadTTTFont a loadTTFont()

## v0.5.7 2022-05-26
- GameObject.py
    - Se crean todas las rutinas de manejo de eventos, los objetos deberán sobreescribirlas

- LittleGameEngine.py
    - Se renombra collidesWithGObjects() a collidesWith()
    - Se ajusta main loop acorde a los cambios de GameObject
    - Se reemplazan vrias list comprehension por ciclos anidados
     
- Se ajustan demos acorde a los cambios

## v0.5.6.1 2022-05-18
- Ajustes a demos

## 2022-05-17
- Se modifica Rectangle.py para trabajar coordenadas (x,y) como floats
- Se modifica Camera.py para manejar su posición siempre con enteros
- Se modifican algunos demos según lo anterior

## 2022-05-15
- Se modifican los demos para no utilizar onCollision() ya que ralentiza demasiado el mainloop

## v0.5.6 2022-05-10
- GameObject.py
    - Agrega método getLayer()
- Sprite.py
    - Corrige método SetImage
- LittleGameEngine.py
    - Agrega método findGObjectsByTag()
- Agrega demo "Cementerio"

## 2022-04-26
- Sprite.py:Corrige errorn en retorno de setImage()

## v0.5.5 2022-04-24
- GameObject.py
    - Permite múltiples rectángulos como colisionador
    - Renombra useColliders() a enableCollider()
    - Agrega setCollider() para establecer varios rectángulos como parte del colisionador
    - Agrega getCollider() para retornar los rectángulos ajustados a las coordenadas
    - Agrega collidesWith() para determinar su colisiona con un GameObject dado
- Sprite.py
    - Agrega getImagesIndex() para retornar el indice actual dentro de la secuencia de imagenes
    - getImagesName() retorna ahora el nombre de la secuencia de imágenes en uso
    - nextImage() retorna ahora el indice actual dentro de la secuencia de imagenes
    - setImage() retorna ahora el indice actual dentro de la secuencia de imagenes
    - Ajusta uso de colisiones
- LittleGameEngine.py
    - Se modifica para manejar múltiples rectángulos en las colisiones
- Se ajustan los demos acorde a los cambios

## v0.5.4 2022-04-23
- Sprite.py: 
    - Modifica constructor para recibir una única referencia a una secuencia de imágenes
    - Cambia de nombre método getCurrentIName() a getImagesName()
    - Elimina método getCurrentIdx() 
    - Cambia de nombre método nextShape() a nextShapeImage()
    - Cambia de nombre y modifica método setShape() por setImage()
- Se ajustan los demos acorde a los cambios

## v0.5.3 2022-04-22
- Cambia coordenadas a la clásica 2D coincidiendo así con las coordenadas de pantalla (IV cuadrante)

## v0.5.2.1 2022-04-21
- Documentación

## v0.5.2 2022-04-07
- Formateo de código con PyDev
- Agrega clase para manejar clic del mouse
- Mejora manejo de eventos de teclado y mouse

## v0.5.1 2020-04-05
- Convierte nombres de variables y métodos a camelCase

## 2022-05-04
- Deja disponible el objeto LGE dentro de cada GameObject
- Ajusta demos

## v0.5.0 2022-04-03
- Corrige métodos DrawCircle y DrawCanvas en la referencia a "surface"
- Modifica activación de eventos para ser realizada sólo sobre un GameObject
- Actualiza demos

## 2022-04-01
- Elimina archivo Clases.md dado que se documentará todo con pdoc3
- Elimina directorio plantilla
- Inicia documentación del código utilizando docstrings
- Corrige método Intersect de la clase Rectangle
- Mejora captura de eventos
- Mejora detección de colisiones
- Corrige seteo de volumen en clips de audio
- AJusta varios demos


## 2022-04-01
- Mueve directorios fonts, images y sounds a subdirectorio resources
- Cambia de nombre archivo Engine.py a LittleGameEngine, reordena su código y cambia invocación inicial a LGE
- Formatea código acorde a pep8
- Modifica código para alinearlo acorde al avance de la versión java JLittleGameMachine
- Ajusta demos acorde a los cambos

## @v0.4.3 2022-03-08
- Agrega eventos OnQuit() a ser ejecutados justo antes de pygame.quit()
- En Canvas elimina método DrawImage() y agrega DrawSurface()
- Agrega demo para mostrar imágenes capturadas con opencv dentro de un Canvas

## @v0.4.2 2022-03-07
- Modifica orde de invocación de los OnCollision()
- Ajusta todas las posiciones y tamaños a int()
- Modifica Engine para que el target de la cámara se ajuste después de todos los eventos y justo antes del rendering
- Agrega plantilla base para generar juegso tipo plataforma
- Cambia todos los archivos de CRLF a LF

## @v0.4.1 2022-03-06
- Agrega método para activar/desactivar eventos en el loop
- Ordena demos/Benchmark
- Ajusta demos acorde al cambio en el loop

## @v0.4.0 2022-03-05
- Optimizaciones en la clase Rectangle
- Optimización en el manejo de colisiones
- Agrega prefijo "_" a las variables de clase
- Agrega método DrawImage en Canvas
- Optimizaciones varias en Engine

## @v0.3.0 2022-02-26
- Se reversa el uso de múltiples colliders en un GameObject hasta no tener un mejor algoritmo
- Optimizaciones varias
- Se ajustan los demos acorde a los cambios


## 2022-02-25
- Renombra Rect.py a Rectangle.py y actualiza todos los archivos afectados
- Agrega demo con pseudo-física
- Modifica uso de colliders permitiendo multiples rectángulos

## 2022-02-24
- Agrego demo de Canvas con rebote de pelotas
- Optimiza CollideRect en Rect
- Avanza en Canvas
- Corrige y mejora demos

## 2022-02-23
- Corrige algunos cálculos en Rect.py y Camera.py
- Los GameObject deben ahora habilitar el uso del collider

demos/Pong/Game.py
from lge.LittleGameEngine import LittleGameEngine
from lge.Canvas import Canvas

from Ball import Ball


class Pong():

    def __init__(self):
        # creamos el juego
        winSize = (640, 640)

        self.lge = LittleGameEngine(winSize, "Pong", (0, 0, 0))
        self.lge.setOnMainUpdate(self.onMainUpdate)
        # self.lge.showColliders((255, 0, 0))

        # cargamos los recursos que usaremos
        resourceDir = "../resources"

        self.lge.loadTTFont("monospace.16", resourceDir + "/fonts/FreeMono.ttf", 16)

        # agregamos la barra de info
        infobar = Canvas((0, 0), (640, 20), "infobar")
        self.lge.addGObjectGUI(infobar)

        # el campo de juego
        field = Canvas((24, 80), (592, 526), "field")
        field.fill((0, 0, 100))
        self.lge.addGObject(field, 0)

        # los bordes
        wall = Canvas((0, 76), (640, 4))
        wall.fill((255, 255, 255))
        wall.setTag("wall-horizontal")
        wall.enableCollider(True)
        self.lge.addGObject(wall, 1)

        wall = Canvas((0, 606), (640, 4))
        wall.fill((255, 255, 255))
        wall.setTag("wall-horizontal")
        wall.enableCollider(True)
        self.lge.addGObject(wall, 1)

        wall = Canvas((20, 80), (4, 526))
        wall.fill((255, 255, 255))
        wall.setTag("wall-vertical")
        wall.enableCollider(True)
        self.lge.addGObject(wall, 1)

        wall = Canvas((616, 80), (4, 526))
        wall.fill((255, 255, 255))
        wall.setTag("wall-vertical")
        wall.enableCollider(True)
        self.lge.addGObject(wall, 1)

        # los actores
        ball = Ball((320, 400), (8, 8), "ball")
        self.lge.addGObject(ball, 1)

        paddle = Canvas((90, 270), (8, 60), "user-paddle")
        paddle.fill((255, 255, 255))
        paddle.setTag("paddle")
        paddle.enableCollider(True)
        paddle.setBounds(field.getRectangle())
        self.lge.addGObject(paddle, 1)

        paddle = Canvas((540, 270), (8, 60), "system-paddle")
        paddle.fill((255, 255, 255))
        paddle.setTag("paddle")
        paddle.enableCollider(True)
        paddle.setBounds(field.getRectangle())
        self.lge.addGObject(paddle, 1)

        self.paddleSpeed = 240

    def onMainUpdate(self, dt):
        # abortamos con la tecla Escape
        if(self.lge.keyPressed(LittleGameEngine.CONSTANTS.K_ESCAPE)):
            self.lge.quit()

        # mostramos info
        mx, my = self.lge.getMousePosition()
        mb1, mb2, mb3 = self.lge.getMouseButtons()

        info = "FPS: %07.2f - gObjs: %03d - Mouse: (%3d,%3d) (%d,%d,%d)" % (
            self.lge.getFPS(),
            self.lge.getCountGObjects(), mx, my,
            mb1, mb2, mb3
        )
        infobar = self.lge.getGObject("infobar")
        infobar.fill((80, 80, 80, 200))
        infobar.drawText(info, (50, 0), "monospace.16", (255, 255, 255))

        # user paddle
        userPaddle = self.lge.getGObject("user-paddle")
        speed = self.paddleSpeed * dt
        x = userPaddle.getX()
        y = userPaddle.getY()

        if (self.lge.keyPressed(self.lge.CONSTANTS.K_UP)):
            userPaddle.setPosition(x, y - speed)
        elif (self.lge.keyPressed(self.lge.CONSTANTS.K_DOWN)):
            userPaddle.setPosition(x, y + speed)

        # la pelota
        ball = self.lge.getGObject("ball")
        # bx = ball.getX()
        by = ball.getY()

        # system paddle
        systemPaddle = self.lge.getGObject("system-paddle")
        px = systemPaddle.getX()
        py = systemPaddle.getY()
        # pw = systemPaddle.getWidth()
        ph = systemPaddle.getHeight()

        if (py + ph / 2 < by):
            py = py + speed
        elif (py + ph / 2 > by):
            py = py - speed
        systemPaddle.setPosition(px, py)

    # main loop
    def run(self, fps):
        self.lge.run(fps)


# ----
game = Pong()
game.run(60)
